base-notebook/start.sh

```diff
@@ -88,18 +88,18 @@ if [ $(id -u) == 0 ] ; then
 
     # Enable sudo if requested
     if [[ "$GRANT_SUDO" == "1" || "$GRANT_SUDO" == 'yes' ]]; then
-        echo "Granting $NB_USER sudo access and appending $CONDA_DIR/bin to sudo PATH"
+        echo "Granting $NB_USER sudo access and prepending $CONDA_DIR/bin to sudo PATH"
         echo "$NB_USER ALL=(ALL) NOPASSWD:ALL" > /etc/sudoers.d/notebook
     fi
 
     # Add $CONDA_DIR/bin to sudo secure_path
-    sed -r "s#Defaults\s+secure_path\s*=\s*\"?([^\"]+)\"?#Defaults secure_path=\"\1:$CONDA_DIR/bin\"#" /etc/sudoers | grep secure_path > /etc/sudoers.d/path
+    sed -r "s#Defaults\s+secure_path\s*=\s*\"?([^\"]+)\"?#Defaults secure_path=\"$CONDA_DIR/bin:\1\"#" /etc/sudoers | grep secure_path > /etc/sudoers.d/path
 
-    # Exec the command as NB_USER with the PATH and the rest of
+    # Exec the command as NB_USER with the PATH from secure_path and the rest of
     # the environment preserved
     run-hooks /usr/local/bin/before-notebook.d
     echo "Executing the command: ${cmd[@]}"
-    exec sudo -E -H -u $NB_USER PATH=$PATH XDG_CACHE_HOME=/home/$NB_USER/.cache PYTHONPATH=${PYTHONPATH:-} "${cmd[@]}"
+    exec sudo -E -H -u $NB_USER XDG_CACHE_HOME=/home/$NB_USER/.cache PYTHONPATH=${PYTHONPATH:-} "${cmd[@]}"
```

Comment:
I wonder about removing `PATH=$PATH` here, what is the motivation for doing so? I'm quite confused in general about this.

https://github.com/jupyter/docker-stacks/pull/1052/files#diff-41f90d7afcdae13f8516195e078d0a203972b5c5105851eaecfc0a98e9739107R147-R185

Reply:
Trying things out locally on my computer, I note:
- using `-E` / `--preserve-env` doesn't preserve PATH, but setting it explicitly with `PATH=$PATH` works, and explicitly setting it to be preserved with `--preserve-env=PATH` works.

I was thinking perhaps this was removed because it served no purpose, but perhaps it does? I'm generally confused.

Should this be part of this PR?

Reply:
@consideRatio thanks for reminding me of this, I had forgotten. It's really painful to come back to something like this a whole year later.

I even wrote [here](https://github.com/jupyter/docker-stacks/issues/1053#issuecomment-605349169) that I like your solution better. I'll see if I can dig in and figure out a reasonable solution...

Reply:
I think I have an idea about what's going on and why we're so confused, although I need to do further experiments to be sure...

There are really two separate things going on. First `sudo` executes `"${cmd[@]}"`, and then this execution inherits environment variables. Often `"${cmd[@]}"` will be `bash`, which will then resolve paths according to the `$PATH` passed to it. However, `sudo` does its own path resolution, independently of `$PATH`. For example, in #1053, `sudo` runs `python`, which `sudo` resolves as `/usr/bin/python` according to `secure_path`. However, if `sudo` runs `bash` and then you type `python` into the shell, then `bash` resolves `python` according to `$PATH`.

I wonder if your #1052 solves #1053... I need to test it, but I'm optimistic.